### PR TITLE
[V1] Prototype Fully Async Detokenizer

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
### SUMMARY:
* In current v1 prototype, we create `RequestOutput` objects in the `LLMEngine` based on the value of `step()` and send them back to the Client. This has implications:
     * The `Detokenizer` runs in a separate process, `step()` may return an empty list as the `Detokenizer` can fall behind. This means that we need to maintain state about how many steps behind each request is (`num_lagged_steps` object), which complicates the code
     * Creating `RequestOutputs` and sending the `RequestOutputs` back to the `MQLLMEngine` run in the main `LLMEngine` process, increasing contention.
* PR makes the `Detokenizer` fully async by using the `Detokenizer` as the message sender (sends the `RequestOutputs` to the `MQLLMEngineClient` --> this avoids needing to send the outputs of the `Detokenizer` back to the `LLMEngine` (reducing the complexity of the `LLMEngine` and reducing overhead).

### ARCH SUMMARY:
image

### TODO:
- [ ] Get agreement from group that this strategy makes sense
- [ ] Figure out how to make it work with `LLM`
- [ ] Guarantee there is no race condition for sending a new request vs sending new tokens
- [ ] Figure out if it is okay for the `MQLLMEngine` and `Detokenizer` to `connect` and `PUSH` to the same socket.
- [ ] Propogate abort to the `Detokenizer`
- [ ] Shutdown for `Detokenizer`
- [ ] Error handling for `Detokenizer`
- [ ] Decide what to do with `AsyncLLMEngine`
- [ ] Optimize the message passing `copy()` like we did for the `MQLLMEngine`
- [ ] Performance benchmark

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/v1/engine/llm_engine.py [hunk 1]
```diff
@@ -300,155 +290,62 @@ def add_request(
...
+                stop_reason=req.stop_reason) for req, num_tokens in sampled
+        ]
+
+        self.detokenizer.send(DetokenizerInputs(data=input_data))
 
     def terminate_detokenizer(self) -> None:
         self.detokenizer.terminate()
 
-    def _make_request_output(
-        self,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/llm_engine.py:389]: we now create the `RequestOutputs` in the `Detokenizer`

File: vllm/v1/engine/llm_engine.py [hunk 2]
```diff
@@ -300,155 +290,62 @@ def add_request(
...
-    def step(self) -> List[RequestOutput]:
-        # NOTE(woosuk): This method may return an empty list when the
-        # detokenizer is still processing the outputs. This should not be
-        # considered as the end of the generation process.
-        # FIXME(woosuk): Currently, the step method is inefficient because it
-        # creates RequestOutput objects for all running requests, while they
-        # may not be needed unless the output is streamed to the client.
+        return self.scheduler.has_unfinished_requests()
+
+    def step(self) -> None:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/llm_engine.py:304]: This technically breaks the API, as `step` no longer returns `RequestOutput`.

Note: this is already how we use the `LLMEngine` in the `MQLLMEngine` with `async_process_outputs`.

File: vllm/v1/engine/llm_engine.py [hunk 3]
```diff
@@ -142,18 +145,6 @@ def __init__(
...
-        self.requests: Dict[str, Request] = {}
-        # NOTE(woosuk): Now that the detokenizer works asynchronously, we need
-        # to keep track of how many steps each request has been lagged behind
-        # in terms of detokenization.
-        # Request id -> how many detokenizer steps the request should wait for.
-        self.num_lagged_steps: Dict[str, int] = {}
-        # OPTIMIZATION: Cache the request output and update it incrementally.
-        # This is used to avoid creating a new RequestOutput object every step.
-        # Request id -> RequestOutput
-        self.request_outputs: Dict[str, RequestOutput] = {}
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/llm_engine.py:155]: The key benefit of this PR is that we simplify the `LLMEngine` by no longer needing to keep track of these data.

File: vllm/engine/multiprocessing/engine.py [hunk 4]
```diff
@@ -95,8 +100,10 @@ def __init__(self,
         self.input_socket.bind(f"{ipc_path}{IPC_INPUT_EXT}")
 
         # Send output stream back to client.
+        # TODO(robertgshaw2): this currently uses the same path as
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/multiprocessing/engine.py:103]: I think this should be okay, but I want to double check. If this is a problem, we can convert this to a separate "error" socket and poll on the error and output socket from the `MQLLMEngineClient`

**Comment by njhill** [comment]
Thanks @robertgshaw2-neuralmagic! Actually @WoosukKwon and I were discussing something quite similar to this last week and I had started writing up a similar proposal ... I'd just been thinking some more about the best way to structure it to retain backwards compatibility with the various existing interfaces.

Here is what I propose:
- We have the core step logic in an `LLMEngine._step` method which returns only token ids in a compact form
- The `LLMEngine.step` method can wrap this and perform the detokenization and stop string evaluation inline, returning the same `List[RequestOutput]` for backwards compatibility in case folks are using this directly (though we would discourage it)
- We rework the `LLM` class to actually start the LLMEngine in a separate process which runs a loop invoking the `_step` method, connected with zmq, basically the same thing as MQLLMEngine. In this case tokenization and detokenization are done on the client side.
- We add a new `AsyncLLM` class which is the same thing just with an async interface. Again pretty much the same as MQLLMEngine except that we do the tokenization/detokenization on the client size, and the messages that we send internally are more compact (not `RequestOutput` objects).

So the main difference is that we don't have a dedicated process for detokenization, we just do it on the client side. The important thing is that we isolate the critical loop in its own process. And we don't have an explicit `MQLLMEngine`, this just becomes the standard behaviour when you use the `LLM` or `AsyncLLM` interfaces.

WDYT?

**Comment by robertgshaw2-neuralmagic** [comment]
> Thanks @robertgshaw2-neuralmagic! Actually @WoosukKwon and I were discussing something quite similar to this last week and I had started writing up a similar proposal ... I'd just been thinking some more about the best way to structure it to retain backwards compatibility with the various existing interfaces.
> 
> Here is what I propose:
> 
> * We have the core step logic in an `LLMEngine._step` method which returns only token ids in a compact form
> * The `LLMEngine.step` method can wrap this and perform the detokenization and stop string evaluation inline, returning the same `List[RequestOutput]` for backwards compatibility in case folks are using this directly (though we would discourage it)
> * We rework the `LLM` class to actually start the LLMEngine in a separate process which runs a loop invoking the `_step` method, connected with zmq, basically the same thing as MQLLMEngine. In this case tokenization and detokenization are done on the client side.
> * We add a new `AsyncLLM` class which is the same thing just with an async interface. Again pretty much the same as MQLLMEngine except that we do the tokenization/detokenization on the client size, and the messages that we send internally are more compact (not `RequestOutput` objects).
> 
> So the main difference is that we don't have a dedicated process for detokenization, we just do it on the client side. The important thing is that we isolate the critical loop in its own process. And we don't have an explicit `MQLLMEngine`, this just becomes the standard behaviour when you use the `LLM` or `AsyncLLM` interfaces.
> 
> WDYT?

I agree with the direction you described

**Comment by robertgshaw2-neuralmagic** [comment]
Closing in favor of #9741 9741

**Pull request closed without merging**